File: pkg/client/supplychain_test.go
package client_test

import (
	"bytes"
	"context"
	"io"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/candlekeep/zot-artifact-store/pkg/client"
	"github.com/candlekeep/zot-artifact-store/test"
)

func TestSupplyChainOperations(t *testing.T) {
	t.Run("Sign artifact", func(t *testing.T) {
		// Given: Test server
		var receivedBody []byte

		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			test.AssertEqual(t, "POST", r.Method, "HTTP method")
			test.AssertEqual(t, "/supplychain/sign/test-bucket/test-key", r.URL.Path, "request path")
			receivedBody, _ = io.ReadAll(r.Body)
			w.WriteHeader(http.StatusOK)
			w.Write([]byte(`{
				"id": "sig-1",
				"bucket": "test-bucket",
				"key": "test-key",
				"signature": "base64-signature",
				"publicKey": "base64-public-key",
				"algorithm": "RSA-SHA256",
				"signedAt": "2024-01-01T00:00:00Z",
				"artifactSha": "sha256-hash"
			}`))
		}))
		defer server.Close()

		c, _ := client.NewClient(&client.Config{BaseURL: server.URL})

		// When: Signing an artifact
		ctx := context.Background()
		sig, err := c.SignArtifact(ctx, "test-bucket", "test-key", "private-key-pem")

		// Then: Signature is created
		test.AssertNoError(t, err, "sign artifact")
		test.AssertTrue(t, sig != nil, "signature should not be nil")
		test.AssertEqual(t, "sig-1", sig.ID, "signature ID")
		test.AssertEqual(t, "RSA-SHA256", sig.Algorithm, "algorithm")
		test.AssertTrue(t, len(receivedBody) > 0, "should send request body")
	})

	t.Run("Get signatures", func(t *testing.T) {
		// Given: Test server with signatures
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			test.AssertEqual(t, "GET", r.Method, "HTTP method")
			test.AssertEqual(t, "/supplychain/signatures/test-bucket/test-key", r.URL.Path, "request path")
			w.WriteHeader(http.StatusOK)
			w.Write([]byte(`{
				"signatures": [
					{
						"id": "sig-1",
						"bucket": "test-bucket",
						"key": "test-key",
						"signature": "signature1",
						"publicKey": "public-key-1",
						"algorithm": "RSA-SHA256",
						"signedAt": "2024-01-01T00:00:00Z",
						"artifactSha": "sha256-hash"
					}
				]
			}`))
		}))
		defer server.Close()

		c, _ := client.NewClient(&client.Config{BaseURL: server.URL})

		// When: Getting signatures
		ctx := context.Background()
		sigs, err := c.GetSignatures(ctx, "test-bucket", "test-key")

		// Then: Returns signature list
		test.AssertNoError(t, err, "get signatures")
		test.AssertTrue(t, len(sigs) == 1, "should have 1 signature")
		test.AssertEqual(t, "sig-1", sigs[0].ID, "signature ID")
	})

	t.Run("Verify signatures", func(t *testing.T) {
		// Given: Test server
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			test.AssertEqual(t, "POST", r.Method, "HTTP method")
			test.AssertEqual(t, "/supplychain/verify/test-bucket/test-key", r.URL.Path, "request path")
			w.WriteHeader(http.StatusOK)
			w.Write([]byte(`{
				"valid": true,
				"signatures": ["sig-1"],
				"message": "All signatures valid",
				"verifiedAt": "2024-01-01T00:00:00Z"
			}`))
		}))
		defer server.Close()

		c, _ := client.NewClient(&client.Config{BaseURL: server.URL})

		// When: Verifying signatures
		ctx := context.Background()
		publicKeys := []string{"public-key-1", "public-key-2"}
		result, err := c.VerifySignatures(ctx, "test-bucket", "test-key", publicKeys)

		// Then: Returns verification result
		test.AssertNoError(t, err, "verify signatures")
		test.AssertTrue(t, result != nil, "result should not be nil")
		test.AssertTrue(t, result.Valid, "should be valid")
		test.AssertTrue(t, len(result.Signatures) == 1, "should have 1 verified signature")
	})

	t.Run("Attach SBOM", func(t *testing.T) {
		// Given: Test server
		sbomContent := "SPDX-2.3 document content"
		var receivedBody []byte

		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			test.AssertEqual(t, "POST", r.Method, "HTTP method")
			test.AssertEqual(t, "/supplychain/sbom/test-bucket/test-key", r.URL.Path, "request path")
			receivedBody, _ = io.ReadAll(r.Body)
			w.WriteHeader(http.StatusOK)
			w.Write([]byte(`{
				"id": "sbom-1",
				"bucket": "test-bucket",
				"key": "test-key",
				"format": "spdx",
				"content": "` + sbomContent + `",
				"createdAt": "2024-01-01T00:00:00Z"
			}`))
		}))
		defer server.Close()

		c, _ := client.NewClient(&client.Config{BaseURL: server.URL})

		// When: Attaching SBOM
		ctx := context.Background()
		sbom, err := c.AttachSBOM(ctx, "test-bucket", "test-key", "spdx", sbomContent)

		// Then: SBOM is attached
		test.AssertNoError(t, err, "attach SBOM")
		test.AssertTrue(t, sbom != nil, "SBOM should not be nil")
		test.AssertEqual(t, "sbom-1", sbom.ID, "SBOM ID")
		test.AssertEqual(t, "spdx", sbom.Format, "SBOM format")
		test.AssertTrue(t, len(receivedBody) > 0, "should send request body")
	})

	t.Run("Get SBOM", func(t *testing.T) {
		// Given: Test server with SBOM
		sbomContent := "SPDX-2.3 document content"

		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			test.AssertEqual(t, "GET", r.Method, "HTTP method")
			test.AssertEqual(t, "/supplychain/sbom/test-bucket/test-key", r.URL.Path, "request path")
			w.WriteHeader(http.StatusOK)
			w.Write([]byte(`{
				"id": "sbom-1",
				"bucket": "test-bucket",
				"key": "test-key",
				"format": "spdx",
				"content": "` + sbomContent + `",
				"createdAt": "2024-01-01T00:00:00Z"
			}`))
		}))
		defer server.Close()

		c, _ := client.NewClient(&client.Config{BaseURL: server.URL})

		// When: Getting SBOM
		ctx := context.Background()
		sbom, err := c.GetSBOM(ctx, "test-bucket", "test-key")

		// Then: Returns SBOM
		test.AssertNoError(t, err, "get SBOM")
		test.AssertTrue(t, sbom != nil, "SBOM should not be nil")
		test.AssertEqual(t, "sbom-1", sbom.ID, "SBOM ID")
		test.AssertEqual(t, "spdx", sbom.Format, "SBOM format")
	})

	t.Run("Add attestation", func(t *testing.T) {
		// Given: Test server
		var receivedBody []byte

		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			test.AssertEqual(t, "POST", r.Method, "HTTP method")
			test.AssertEqual(t, "/supplychain/attestations/test-bucket/test-key", r.URL.Path, "request path")
			receivedBody, _ = io.ReadAll(r.Body)
			w.WriteHeader(http.StatusOK)
			w.Write([]byte(`{
				"id": "att-1",
				"bucket": "test-bucket",
				"key": "test-key",
				"type": "build",
				"data": {"buildId": "123", "status": "success"},
				"createdAt": "2024-01-01T00:00:00Z",
				"createdBy": "ci-system"
			}`))
		}))
		defer server.Close()

		c, _ := client.NewClient(&client.Config{BaseURL: server.URL})

		// When: Adding attestation
		ctx := context.Background()
		attData := map[string]interface{}{
			"buildId": "123",
			"status":  "success",
		}
		att, err := c.AddAttestation(ctx, "test-bucket", "test-key", "build", attData)

		// Then: Attestation is added
		test.AssertNoError(t, err, "add attestation")
		test.AssertTrue(t, att != nil, "attestation should not be nil")
		test.AssertEqual(t, "att-1", att.ID, "attestation ID")
		test.AssertEqual(t, "build", att.Type, "attestation type")
		test.AssertTrue(t, bytes.Contains(receivedBody, []byte("build")), "should contain type in body")
	})

	t.Run("Get attestations", func(t *testing.T) {
		// Given: Test server with attestations
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			test.AssertEqual(t, "GET", r.Method, "HTTP method")
			test.AssertEqual(t, "/supplychain/attestations/test-bucket/test-key", r.URL.Path, "request path")
			w.WriteHeader(http.StatusOK)
			w.Write([]byte(`{
				"attestations": [
					{
						"id": "att-1",
						"bucket": "test-bucket",
						"key": "test-key",
						"type": "build",
						"data": {"buildId": "123"},
						"createdAt": "2024-01-01T00:00:00Z"
					},
					{
						"id": "att-2",
						"bucket": "test-bucket",
						"key": "test-key",
						"type": "test",
						"data": {"passed": true},
						"createdAt": "2024-01-01T01:00:00Z"
					}
				]
			}`))
		}))
		defer server.Close()

		c, _ := client.NewClient(&client.Config{BaseURL: server.URL})

		// When: Getting attestations
		ctx := context.Background()
		atts, err := c.GetAttestations(ctx, "test-bucket", "test-key")

		// Then: Returns attestation list
		test.AssertNoError(t, err, "get attestations")
		test.AssertTrue(t, len(atts) == 2, "should have 2 attestations")
		test.AssertEqual(t, "att-1", atts[0].ID, "first attestation ID")
		test.AssertEqual(t, "build", atts[0].Type, "first attestation type")
		test.AssertEqual(t, "att-2", atts[1].ID, "second attestation ID")
		test.AssertEqual(t, "test", atts[1].Type, "second attestation type")
	})
}

File: pkg/client-js/tests/supplychain.test.ts
/**
 * Tests for supply chain operations
 */

import { Client } from '../src/client';
import { Config } from '../src/types';
import axios from 'axios';

// Mock axios
jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;

describe('Supply Chain Operations', () => {
  let client: Client;
  let mockAxiosInstance: any;

  beforeEach(() => {
    jest.clearAllMocks();

    mockAxiosInstance = {
      get: jest.fn(),
      post: jest.fn(),
      put: jest.fn(),
      delete: jest.fn(),
      head: jest.fn(),
      defaults: {
        headers: {
          common: {},
        },
      },
      interceptors: {
        response: {
          use: jest.fn(),
        },
      },
    };

    mockedAxios.create.mockReturnValue(mockAxiosInstance);

    const config: Config = {
      baseURL: 'https://test.example.com',
      token: 'test-token',
    };
    client = new Client(config);
  });

  it('should sign artifact', async () => {
    // Given: Artifact and private key
    mockAxiosInstance.post.mockResolvedValue({
      data: {
        id: 'sig123',
        artifactDigest: 'sha256:abc123',
        signature: 'signature-data',
        algorithm: 'RSA-SHA256',
        signedBy: 'test-signer',
        timestamp: '2024-01-15T10:30:00Z',
      },
    });

    // When: Signing artifact
    const signature = await client.signArtifact('mybucket', 'myfile.tar.gz', 'private-key-pem');

    // Then: Should return signature
    expect(signature.id).toBe('sig123');
    expect(signature.artifactDigest).toBe('sha256:abc123');
    expect(signature.algorithm).toBe('RSA-SHA256');
    expect(signature.signedBy).toBe('test-signer');
    expect(signature.timestamp).toBeInstanceOf(Date);
  });

  it('should get signatures', async () => {
    // Given: Signed artifact
    mockAxiosInstance.get.mockResolvedValue({
      data: {
        signatures: [
          {
            id: 'sig1',
            artifactDigest: 'sha256:abc123',
            signature: 'sig-data-1',
            algorithm: 'RSA-SHA256',
            signedBy: 'signer1',
            timestamp: '2024-01-15T10:30:00Z',
          },
          {
            id: 'sig2',
            artifactDigest: 'sha256:abc123',
            signature: 'sig-data-2',
            algorithm: 'RSA-SHA256',
            signedBy: 'signer2',
            timestamp: '2024-01-15T11:00:00Z',
          },
        ],
      },
    });

    // When: Getting signatures
    const signatures = await client.getSignatures('mybucket', 'myfile.tar.gz');

    // Then: Should return list of signatures
    expect(signatures).toHaveLength(2);
    expect(signatures[0].id).toBe('sig1');
    expect(signatures[1].id).toBe('sig2');
  });

  it('should verify signatures', async () => {
    // Given: Signed artifact and public key
    mockAxiosInstance.post.mockResolvedValue({
      data: {
        valid: true,
        message: 'All signatures valid',
        signatures: [
          {
            id: 'sig1',
            artifactDigest: 'sha256:abc123',
            signature: 'sig-data',
            algorithm: 'RSA-SHA256',
            signedBy: 'signer1',
            timestamp: '2024-01-15T10:30:00Z',
          },
        ],
      },
    });

    // When: Verifying signatures
    const result = await client.verifySignatures('mybucket', 'myfile.tar.gz', ['public-key-pem']);

    // Then: Should return verification result
    expect(result.valid).toBe(true);
    expect(result.message).toBe('All signatures valid');
    expect(result.signatures).toHaveLength(1);
  });

  it('should verify signatures failure', async () => {
    // Given: Artifact with invalid signature
    mockAxiosInstance.post.mockResolvedValue({
      data: {
        valid: false,
        message: 'Signature verification failed',
        signatures: [],
      },
    });

    // When: Verifying signatures
    const result = await client.verifySignatures('mybucket', 'myfile.tar.gz', ['public-key-pem']);

    // Then: Should return failed result
    expect(result.valid).toBe(false);
    expect(result.message).toContain('failed');
  });

  it('should attach SBOM', async () => {
    // Given: Artifact and SBOM content
    mockAxiosInstance.post.mockResolvedValue({
      data: {
        id: 'sbom123',
        artifactDigest: 'sha256:abc123',
        format: 'spdx',
        content: '{"spdxVersion": "2.3"}',
        timestamp: '2024-01-15T10:30:00Z',
      },
    });

    // When: Attaching SBOM
    const sbom = await client.attachSBOM(
      'mybucket',
      'myfile.tar.gz',
      'spdx',
      '{"spdxVersion": "2.3"}'
    );

    // Then: Should return SBOM
    expect(sbom.id).toBe('sbom123');
    expect(sbom.format).toBe('spdx');
    expect(sbom.artifactDigest).toBe('sha256:abc123');
    expect(sbom.timestamp).toBeInstanceOf(Date);
  });

  it('should get SBOM', async () => {
    // Given: Artifact with SBOM
    mockAxiosInstance.get.mockResolvedValue({
      data: {
        id: 'sbom123',
        artifactDigest: 'sha256:abc123',
        format: 'spdx',
        content: '{"spdxVersion": "2.3", "packages": []}',
        timestamp: '2024-01-15T10:30:00Z',
      },
    });

    // When: Getting SBOM
    const sbom = await client.getSBOM('mybucket', 'myfile.tar.gz');

    // Then: Should return SBOM
    expect(sbom.id).toBe('sbom123');
    expect(sbom.format).toBe('spdx');
    expect(sbom.content).toContain('spdxVersion');
  });

  it('should add attestation', async () => {
    // Given: Artifact and attestation data
    mockAxiosInstance.post.mockResolvedValue({
      data: {
        id: 'att123',
        artifactDigest: 'sha256:abc123',
        type: 'build',
        data: {
          buildId: '12345',
          status: 'success',
          tests: 142,
        },
        timestamp: '2024-01-15T10:30:00Z',
      },
    });

    // When: Adding attestation
    const attestation = await client.addAttestation('mybucket', 'myfile.tar.gz', 'build', {
      buildId: '12345',
      status: 'success',
      tests: 142,
    });

    // Then: Should return attestation
    expect(attestation.id).toBe('att123');
    expect(attestation.type).toBe('build');
    expect(attestation.data.buildId).toBe('12345');
    expect(attestation.data.status).toBe('success');
  });

  it('should get attestations', async () => {
    // Given: Artifact with attestations
    mockAxiosInstance.get.mockResolvedValue({
      data: {
        attestations: [
          {
            id: 'att1',
            artifactDigest: 'sha256:abc123',
            type: 'build',
            data: { buildId: '123', status: 'success' },
            timestamp: '2024-01-15T10:30:00Z',
          },
          {
            id: 'att2',
            artifactDigest: 'sha256:abc123',
            type: 'test',
            data: { testsPassed: 142, testsFailed: 0 },
            timestamp: '2024-01-15T11:00:00Z',
          },
        ],
      },
    });

    // When: Getting attestations
    const attestations = await client.getAttestations('mybucket', 'myfile.tar.gz');

    // Then: Should return attestations
    expect(attestations).toHaveLength(2);
    expect(attestations[0].type).toBe('build');
    expect(attestations[1].type).toBe('test');
    expect(attestations[0].data.buildId).toBe('123');
    expect(attestations[1].data.testsPassed).toBe(142);
  });
});
